[AArch64] Correct SCVTF/UCVTF instructions for vector input

Amichaxx opened the pull request:
This pull request improves support for scalar floating-point conversions from integer vectors on AArch64, specifically for the `scvtf` and `ucvtf` instructions. It fixes pattern matching so that single-element conversions from vectors now generate the expected scalar instructions and adds a new test to verify correct behavior for extracting a lane from a widened vector.

**Pattern matching and code generation improvements:**
* Added new patterns in `AArch64InstrInfo.td` to correctly match conversions from `v2i32` to `v1f64` using `scvtf` and `ucvtf`, ensuring the scalar instructions (`scvtf d0, s0` and `ucvtf d0, s0`) are generated when extracting a single lane.

**Test updates and additions:**
* Updated `scvtf_f64i32_simple` and `ucvtf_f64i32_simple` tests in `fprcvt-cvtf.ll` to reflect the correct generation of scalar instructions, removing previous comments about incorrect codegen and showing the expected output. 
* Added a new test `uitofp_sext_v2i32_extract_lane0` to verify correct code generation when extracting a lane from a widened vector and converting to double.

Amichaxx committed:
[AArch64] Correct SCVTF/UCVTF instructions for vector inputs

CarolineConcatto reviewed (approved):
Thank you Amina,

Align before merging

efriedma-quic commented on llvm/lib/Target/AArch64/AArch64InstrFormats.td:
```diff
@@ -5520,6 +5520,11 @@ multiclass IntegerToFPSIMDScalar<bits<2> rmode, bits<3> opcode, string asm, SDPa
     let Inst{23-22} = 0b00; // 32-bit FPR flag
   }
+  def : Pat<(v1f64 (extract_subvector (v2f64 (node (v2i64 (sext (v2i32 V64:$Rn))))), (i64 0))),
+        (!cast<Instruction>(NAME # DSr) (EXTRACT_SUBREG V64:$Rn, ssub))>;
```
This will match `uitofp(sext(...))`, which is not a valid transform.

Amichaxx committed:
Add isel patterns for SCVTF/UCVTF

Amended patterns to match the signed path (sitofp(sext(...))) to SCVTFDSr and
the unsigned path (uitofp(zext(...))) to UCVTFDSr. Moved to AArch64InstrInfo.td.
Added test in fprcvt-cvtf.ll to verify.

Amichaxx commented on llvm/lib/Target/AArch64/AArch64InstrFormats.td:
```diff
@@ -5520,6 +5520,11 @@ multiclass IntegerToFPSIMDScalar<bits<2> rmode, bits<3> opcode, string asm, SDPa
+        (!cast<Instruction>(NAME # DSr) (EXTRACT_SUBREG V64:$Rn, ssub))>;
+
+  def : Pat<(v1f64 (extract_subvector (v2f64 (node (v2i64 (zext (v2i32 V64:$Rn))))), (i64 0))),
+          (!cast<Instruction>(NAME # DSr) (EXTRACT_SUBREG V64:$Rn, ssub))>;
```
Patterns have since been removed. I've ensured the new ones I've added are aligned.

Amichaxx committed:
Revision

- Removed change to check lines of added test uitofp_sext_v2i32_extract_lane0
- Removed irrelevant whitespace
- Reverted to original comment on scvtf_f32i64_neg for future revision

Amichaxx commented on llvm/lib/Target/AArch64/AArch64InstrFormats.td:
```diff
@@ -5520,7 +5520,6 @@ multiclass IntegerToFPSIMDScalar<bits<2> rmode, bits<3> opcode, string asm, SDPa
     let Inst{31} = 1; // 64-bit FPR flag
     let Inst{23-22} = 0b00; // 32-bit FPR flag
   }
-
```
Removed now.

efriedma-quic reviewed (approved):
LGTM
